tiatoolbox/models/segmentation/semantic_segmentor.py

```diff
@@ -0,0 +1,983 @@
+        ...         {"units": "mpp", "resolution": 0.50},
+        ...     ],
+        ...     patch_input_shape=[2048, 2048],
+        ...     patch_output_shape=[1024, 1024],
+        ...     stride_shape=[512, 512],
+        ...     save_resolution={"units": "mpp", "resolution": 4.0},
+        ... )
+
+    """
+
+    # We predefine to follow enforcement, actual initialization in init
+    patch_size = None
+    input_resolutions = None
+    output_resolutions = None
+
+    def __init__(
+        self,
+        input_resolutions: List[dict],
+        output_resolutions: List[dict],
+        patch_input_shape: Union[List[int], np.ndarray],
+        patch_output_shape: Union[List[int], np.ndarray],
+        save_resolution: dict = None,
+        **kwargs,
+    ):
+        self.patch_input_shape = patch_input_shape
+        self.patch_output_shape = patch_output_shape
+        self.stride_shape = None
+        self.input_resolutions = input_resolutions
+        self.output_resolutions = output_resolutions
+
+        self.resolution_unit = input_resolutions[0]["units"]
+        self.save_resolution = save_resolution
+
+        for variable, value in kwargs.items():
+            self.__setattr__(variable, value)
+
+        self._validate()
+
+        if self.resolution_unit == "mpp":
+            self.highest_input_resolution = min(
+                self.input_resolutions, key=lambda x: x["resolution"]
+            )
+        else:
+            self.highest_input_resolution = max(
+                self.input_resolutions, key=lambda x: x["resolution"]
+            )
+
+    def _validate(self):
+        """Validate the data format."""
+        resolutions = self.input_resolutions + self.output_resolutions
+        units = [v["units"] for v in resolutions]
+        units = np.unique(units)
+        if len(units) != 1 or units[0] not in [
+            "power",
+            "baseline",
+            "mpp",
+        ]:
+            raise ValueError("Invalid resolution units.")
+
+    @staticmethod
+    def scale_to_highest(resolutions: List[dict], unit: str):
+        """Get scaling factor from input resolutions.
+
+        This will convert resolutions to scaling factor with repsect to
+        highest resolutions found in the input list of resolutions.
+
+        Args:
+            resolutions (list): a list of resolutions where each defined
+                as `{'resolution': value, 'unit': value}`
+            unit (str): unit that the the resolutions are at.
+
+        Returns:
+            :class:`numpy.ndarray`: an 1D array of scaling factor having the same
+                length as `resolutions`
+
+        """
+        old_val = [v["resolution"] for v in resolutions]
+        if unit == "baseline":
+            new_val = old_val
+        elif unit == "mpp":
+            new_val = np.min(old_val) / np.array(old_val)
+        else:
+            # when being power
+            new_val = np.array(old_val) / np.max(old_val)
```

Comment:
This will run for any other string. you should check for == power and otherwise raise a value error.